python/lsst/daf/butler/_formatter.py

```diff
@@ -51,12 +63,1034 @@
+                can_be_cached = False
+                if uri != local_uri:
+                    # URI was remote and file was downloaded
+                    cache_msg = ""
+
+                    if cache_manager.should_be_cached(cache_ref):
+                        # In this scenario we want to ask if the downloaded
+                        # file should be cached but we should not cache
+                        # it until after we've used it (to ensure it can't
+                        # be expired whilst we are using it).
+                        can_be_cached = True
+
+                        # Say that it is "likely" to be cached because
+                        # if the formatter read fails we will not be
+                        # caching this file.
+                        cache_msg = " and likely cached"
+
+                    msg = f"(via download to local file{cache_msg})"
+
+                log.debug(
+                    "Reading%s from location %s %s with formatter %s",
+                    f" component {log_component}" if log_component else "",
+                    uri,
+                    msg,
+                    self.name(),
+                )
+
+                with time_this(
+                    log,
+                    msg="Reading%s from location %s %s with formatter %s",
+                    args=(
+                        f" component {log_component}" if log_component else "",
+                        uri,
+                        msg,
+                        self.name(),
+                    ),
+                ):
+                    result = self.read_from_local_file(local_uri, component=component)
+
+                # File was read successfully so can move to cache
+                if can_be_cached:
+                    cache_manager.move_to_cache(local_uri, cache_ref)
+
+        return result
+
+    def read_from_uri(self, uri: ResourcePath, component: str | None = None) -> Any:
+        """Read a dataset from a URI that can be local or remote.
+
+        Parameters
+        ----------
+        uri : `lsst.resources.ResourcePath`
+            URI to use to read the dataset. This URI can be local or remote
+            and can refer to the actual resource or to a locally cached file.
+        component : `str` or `None`, optional
+            The component to be read from the dataset.
+
+        Returns
+        -------
+        in_memory_dataset : `typing.Any`
+            The Python object read from the resource.
+
+        Raises
+        ------
+        FormatterNotImplementedError
+            Raised if there is no support for direct reads from a, possibly,
+            remote URI.
+
+        Notes
+        -----
+        This method is only called if the class property
+        ``can_read_from_uri`` is set to `True`.
+
+        It is possible that a cached local file will be given to this method
+        even if it was originally a remote URI. This can happen if the write
```

Comment:
No. I mean the write. The file can be written to the cache during `butler.put`. The "it" is the URI to the thing being read. It might not be remote even if originally it was a remote URI.